controllers/gatekeepersync/gatekeeper_constraint_sync.go

```diff
@@ -0,0 +1,434 @@
+var _ reconcile.Reconciler = &GatekeeperConstraintReconciler{}
+
+// Used to track sent messages for a particular Gatekeeper constraint.
+type policyKindName struct {
+	Policy string
+	Kind   string
+	Name   string
+}
+
+// GatekeeperConstraintReconciler is responsible for relaying Gatekeeper constraint audit results as policy status
+// events.
+type GatekeeperConstraintReconciler struct {
+	client.Client
+	Scheme             *runtime.Scheme
+	ClusterNamespace   string
+	DynamicClient      dynamic.Interface
+	InstanceName       string
+	ClientSet          *kubernetes.Clientset
+	ConstraintsWatcher depclient.DynamicWatcher
+	// A cache of sent messages to avoid repeating status events due to race conditions. Each value is a SHA1
+	// digest.
+	lastSentMessages sync.Map
+}
+
+func (r *GatekeeperConstraintReconciler) GetInstanceName() string {
+	return r.InstanceName
+}
+
+func (r *GatekeeperConstraintReconciler) GetClusterNamespace() string {
+	return r.ClusterNamespace
+}
+
+func (r *GatekeeperConstraintReconciler) GetClientSet() *kubernetes.Clientset {
+	return r.ClientSet
+}
+
+func (r *GatekeeperConstraintReconciler) GetControllerName() string {
+	return ControllerName
+}
+
+//+kubebuilder:rbac:groups=policy.open-cluster-management.io,resources=policies,verbs=get;list;watch
+//+kubebuilder:rbac:groups=constraints.gatekeeper.sh,resources=*,verbs=create;get;list;watch
+//+kubebuilder:rbac:groups=admissionregistration.k8s.io,resources=validatingwebhookconfigurations,resourceNames=gatekeeper-validating-webhook-configuration,verbs=get;list;watch
+//+kubebuilder:rbac:groups=core,resources=events,verbs=create;delete;get;list;patch;update;watch
+
+// Reconcile handles Policy objects that contain a Gatekeeper constraint and relays status messages from Gatekeeper
+// audit results. Every time a Gatekeeper constraint in a Policy is updated, a reconcile on the Policy is triggered.
+func (r *GatekeeperConstraintReconciler) Reconcile(
+	ctx context.Context, request reconcile.Request,
+) (
+	reconcile.Result, error,
+) {
+	log := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	log.Info("Reconciling a Policy with one or more Gatekeeper constraints")
+
+	policyObjID := depclient.ObjectIdentifier{
+		Group:     policyv1.GroupVersion.Group,
+		Version:   policyv1.GroupVersion.Version,
+		Kind:      "Policy",
+		Namespace: request.Namespace,
+		Name:      request.Name,
+	}
+	policy := &policyv1.Policy{}
+
+	err := r.Get(ctx, request.NamespacedName, policy)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.V(1).Info("The Policy was deleted. Cleaning up watchers and status message cache.")
+
+			r.lastSentMessages.Range(func(key, value any) bool {
+				keyTyped := key.(policyKindName)
+				if keyTyped.Policy == request.Name {
+					r.lastSentMessages.Delete(keyTyped)
+				}
+
+				return true
+			})
+
+			err := r.ConstraintsWatcher.RemoveWatcher(policyObjID)
```

Comment:
I believe this only returns a non-nil error if the policyObjID is invalid... it shouldn't be invalid, so it should always be nil, but if it *was* somehow invalid then the error here would cause the policy to be re-queued forever.